providers/powerdns/powerdnsProvider.go

```diff
@@ -0,0 +1,269 @@
+	api := &PowerDNS{}
+
+	api.APIKey = m["apiKey"]
+	if api.APIKey == "" {
+		return nil, fmt.Errorf("PowerDNS API Key is required")
+	}
+
+	api.APIUrl = m["apiUrl"]
+	if api.APIUrl == "" {
+		return nil, fmt.Errorf("PowerDNS API URL is required")
+	}
+
+	api.ServerName = m["serverName"]
+	if api.ServerName == "" {
+		return nil, fmt.Errorf("PowerDNS server name is required")
+	}
+
+	// load js config
+	if len(metadata) != 0 {
+		err := json.Unmarshal(metadata, api)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var nss []string
+	for _, ns := range api.DefaultNS {
+		nss = append(nss, ns[0:len(ns)-1])
+	}
+	var err error
+	api.nameservers, err = models.ToNameservers(nss)
+	if err != nil {
+		return api, err
+	}
+
+	var clientErr error
+	api.client, clientErr = pdns.New(
+		pdns.WithBaseURL(api.APIUrl),
+		pdns.WithAPIKeyAuthentication(api.APIKey),
+	)
+	return api, clientErr
+}
+
+// GetNameservers returns the nameservers for a domain.
+func (api *PowerDNS) GetNameservers(string) ([]*models.Nameserver, error) {
+	var r []string
+	for _, j := range api.nameservers {
+		r = append(r, j.Name)
+	}
+	return models.ToNameservers(r)
+}
+
+// ListZones returns all the zones in an account
+func (api *PowerDNS) ListZones() ([]string, error) {
+	var result []string
+	zones, err := api.client.Zones().ListZones(context.Background(), api.ServerName)
+	if err != nil {
+		return result, err
+	}
+	for _, zone := range zones {
+		result = append(result, zone.Name)
+	}
+	return result, nil
+}
+
+// GetZoneRecords gets the records of a zone and returns them in RecordConfig format.
+func (api *PowerDNS) GetZoneRecords(domain string) (models.Records, error) {
+	zone, err := api.client.Zones().GetZone(context.Background(), api.ServerName, domain)
+	if err != nil {
+		return nil, err
+	}
+
+	curRecords := models.Records{}
+	// loop over grouped records by type, called RRSet
+	for _, rrset := range zone.ResourceRecordSets {
+		if rrset.Type == "SOA" {
+			continue
+		}
+		// loop over single records of this group and create records
+		for _, pdnsRecord := range rrset.Records {
+			r, err := toRecordConfig(domain, pdnsRecord, rrset.TTL, rrset.Name, rrset.Type)
+			if err != nil {
+				return nil, err
+			}
+			curRecords = append(curRecords, r)
+		}
+	}
+
+	return curRecords, nil
+}
+
+// GetDomainCorrections returns a list of corrections to update a domain.
+func (api *PowerDNS) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	var corrections []*models.Correction
+
+	// record corrections
+	curRecords, err := api.GetZoneRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// post-process records
+	dc.Punycode()
+	models.PostProcessRecords(curRecords)
+
+	// create record diff by group
+	keysToUpdate := (diff.New(dc)).ChangedGroups(curRecords)
+	desiredRecords := dc.Records.GroupedByKey()
+
+	// create corrections by group
+	for label, msgs := range keysToUpdate {
+		labelName := ensureDotSuffix(label.NameFQDN)
```

Comment:
This can simply be: `labelName := label.NameFQDN + "."`

You are guaranteed  NameFQDN will never end in a `.`